services/auth/shortcut.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package auth
+
+import (
+	"net/http"
+
+	user_model "code.gitea.io/gitea/models/user"
+)
+
+// Shortcut is a group of verification methods that is like Group.
+// It tries each method in order and returns the first non-nil user,
+// but it never returns error even if some method returns error.
+// It is useful for some methods that share the same protocol, shortcut can check them first.
+// For example, OAuth2 and conan.Auth both read token from "Authorization: Bearer <token>" header,
+// If OAuth2 returns error, it is possible that the token is for conan.Auth but it has no chance to check.
+// And Shortcut solves this problem by:
+//
+//	NewGroup(
+//	    Shortcut{&OAuth2, &conan.Auth},
+//	    &OAuth2,
+//	    &auth.Basic{},
+//	    &nuget.Auth{},
+//	    &conan.Auth{},
+//	    &chef.Auth{},
+//	)
```

Comment:
I didn't get the point why OAuth2 and conan.Auth duplicate

Reply:
#### Case 1: the token is for `conan.Auth.`

`Shortcut[OAuth2]` return an error, `Shortcut[conan.Auth]` return a user, so `Shortcut` return the user, and stop.

#### Case 2: the token is for `conan.Auth`, but it's wrong

`Shortcut[OAuth2]` return an error, `Shortcut[conan.Auth]` return an error, so `Shortcut` return **nil** and continue.

`OAuth2 ` return an error (a nit, it should be `conan.Auth` to return an error, but the result is correct), and stop.

#### Case 3: the token is for `OAuth2`, but as query string, not in `Authorization` header. And the token is wrong

`Shortcut[OAuth2]` return an error, `Shortcut[conan.Auth]` return nil, so `Shortcut` return **nil** and continue.

`OAuth2 ` return an error and stop.

Reply:
Could it refactor the `Shortcut` to a `AuthGroup`?

* If any method in the group succeeds, use the method
* If any method fails, return the first error
* If none of the methods auths, do nothing

Then no need to duplicate the methods.